SimPEG/potential_fields/gravity/simulation.py

```diff
@@ -27,27 +242,132 @@ class Simulation3DIntegral(BasePFSimulation):
+    sensitivity_dtype : numpy.dtype, optional
+        Data type that will be used to build the sensitivity matrix.
+    store_sensitivities : str
+        Options for storing sensitivity matrix. There are 3 options
+
+        - 'ram': sensitivities are stored in the computer's RAM
+        - 'disk': sensitivities are written to a directory
+        - 'forward_only': you intend only do perform a forward simulation and
+          sensitivities do not need to be stored
+
+    sensitivity_path : str, optional
+        Path to store the sensitivity matrix if ``store_sensitivities`` is set
+        to ``"disk"``. Default to "./sensitivities".
+    engine : str, optional
+       Choose which engine should be used to run the forward model:
+       ``"geoana"`` or "``choclo``".
+    choclo_parallel : bool, optional
+        If True, the simulation will run in parallel. If False, it will
+        run in serial. If ``engine`` is not ``"choclo"`` this argument will be
+        ignored.
     """
 
     rho, rhoMap, rhoDeriv = props.Invertible("Density")
 
-    def __init__(self, mesh, rho=None, rhoMap=None, **kwargs):
+    def __init__(
+        self,
+        mesh,
+        rho=None,
+        rhoMap=None,
+        engine="geoana",
+        choclo_parallel=True,
+        **kwargs,
+    ):
         super().__init__(mesh, **kwargs)
         self.rho = rho
         self.rhoMap = rhoMap
         self._G = None
         self._gtg_diagonal = None
         self.modelMap = self.rhoMap
+        self.choclo_parallel = choclo_parallel
+        self.engine = engine
+        self._sanity_checks_engine(kwargs)
+        # Define jit functions
+        if self.engine == "choclo":
+            if choclo_parallel:
+                self._fill_sensitivity_matrix = _fill_sensitivity_matrix_parallel
+                self._forward_gravity = _forward_gravity_parallel
+            else:
+                self._fill_sensitivity_matrix = _fill_sensitivity_matrix_serial
+                self._forward_gravity = _forward_gravity_serial
+
+    def _sanity_checks_engine(self, kwargs):
+        """
+        Sanity checks for the engine parameter.
+
+        Needs the kwargs passed to the __init__ method to raise some warnings.
+        """
+        if self.engine not in ("choclo", "geoana"):
+            raise ValueError(
+                f"Invalid engine '{self.engine}'. Choose from 'geoana' or 'choclo'."
+            )
+        if self.engine == "choclo" and choclo is None:
+            raise ImportError(
+                "The choclo package couldn't be found."
+                "Running a gravity simulation with 'engine=\"choclo\"' needs "
+                "choclo to be installed."
+                "\nTry installing choclo with:"
+                "\n    pip install choclo"
+                "\nor:"
+                "\n    conda install choclo"
+            )
+        # Warn if n_processes has been passed
+        if self.engine == "choclo" and "n_processes" in kwargs:
```

Comment:
where is `n_processes` being ignored? I might be missing something but I see it still being passed to `BasePFSimulation` and set as normal

Reply:
Thanks @jedman for the review!

You're right, `n_processes` is still being passed to `BasePFSimulation` and set as an attribute. But when using the `choclo` engine, it's not being used at all. `n_processes` is used within the `BasePFSimulation.linear_operator` method, which is not called when using `choclo` as engine.

That being said, I agree it's not 100% clear that `n_processes` is being ignored. Do you think we should remove it from the `kwargs` or set it to `None` after raising the warning?

Reply:
ah, that's the code path I was missing, thanks. Yes, I think it would be clearer to set `n_processes=None` after raising the warning

Reply:
Done!